students/javarb/sessions/2019/may/Session_34-35-Serialization/Session_34-35-Serialization.md

```diff
@@ -0,0 +1,225 @@
+```
+
+In my first approximation  solution. I was using a single class called `Serialization.java` and a POJO called `Person.java`. 
+The idea behind the benchmark was use a single time in miliseconds comparison after serialize a list set of persons. 
+
+Receiving feedback of Roger, I finished to write `main()` method of `Serialization.java`. There, a for loop was populating a `List` of `Person` objects and passing over another for loop to store the benchmark results of calculate the miliseconds time that takes to serialize the list of `Person`s. That results are stored in another `List` that is passes to `mean` method in order to calculate mean value and print it:
+```java
+//...
+List<Person> persons = new ArrayList<>();
+
+for (int i = 0; i < 20000; i++) {
+	Person person = new Person();
+    person.setId(i);
+    person.setName("Name_" + i);
+    person.setAge(i % 100);
+    persons.add(i, person);
+}
+
+List<Long> gsonResults = new ArrayList<>();
+List<Long> jacksonResults = new ArrayList<>();
+List<Long> fastjsonResults = new ArrayList<>();
+
+for (int i = 0; i < 150; i++) {
+	gsonResults.add(benchmarkGson(persons));
+	jacksonResults.add(benchmarkJackson(persons));
+	fastjsonResults.add(benchmarkFastjson(persons));
+}
+
+long gsonMean = mean(gsonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+
+long jacksonMean = mean(jacksonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+
+long fastjsonMean = mean(fastjsonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+//...
+```
+
+Following code chunk, performs serialization returning miliseconds time benchmark in this example for Gson serialization library:
+```java
+private static long benchmarkGson(List<Person> persons) {
+	long begin = System.currentTimeMillis();
+	new Gson().toJson(persons);
+	return System.currentTimeMillis() - begin;
+}
+```
+
+Finally, method on charge of return the mean can be seen following:
+```java
+private long mean(List<Long> benchmarks) {
+	long sum = 0;  
+	for (Long benchmark : benchmarks) {
+		sum += benchmark;
+	}
+	return sum / benchmarks.size();
+}
+```
+
+A for loop was used in order to add up benchmarks in order to after get the final mean time benchmarked. 
+
+**Note:** As a side note, there is another way to iterate by using `foreach()` lambda funcion:
+```java
+benchmarks.forEach(b -> sum +=b);
+```
+As can be seen, this kind of function is practical to use but isn't possible to modify the elements we are iterating over, so we cannot use it here.
+
+
+#### Part 2 - Benchmark project in OOP
+
+After code was running, we saw code could be improved through asimilation of OOP concepts which allow us to include any library we want to our benchmark without need to modify our main funcions and code. So instead of that, we could add another serialization library just by including the library on `build.gradle`, create the corresponding class for that implementation,  and instantiating/ executing library's `benchmark()` method.
```

Comment:
```suggestion
After the code was running, we saw that the code could be improved through assimilation of OOP concepts which allow us to include any library we want to our benchmark without the need to modify our main functions and code. So instead of that, we could add another serialization library just by including the library in `build.gradle`, create the corresponding class for that implementation, and instantiating/ executing the library's `benchmark()` method.
```